src/Libraries/DSCPython/DSCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,428 @@
+using System.Text.RegularExpressions;
+using PythonEngine = Python.Runtime.PythonEngine;
+
+namespace DSCPython
+{
+    internal class DSCPythonCodeCompletionProviderCore : PythonCodeCompletionProviderCommon, ILogSource, IDisposable
+    {
+        #region Private members
+
+        /// <summary>
+        /// The engine used for autocompletion.  This essentially keeps
+        /// track of the state of the editor, allowing access to variable types and
+        /// imported symbols.
+        /// </summary>
+        private PyScope scope;
+
+        private string uniquePythonScopeName = "uniqueScope";
+        private static string globalScopeName = "cPythonCompletionGlobal";
+        /// <summary>
+        /// The scope used by the engine.  This is where all the loaded symbols
+        /// are stored.  It's essentially an environment dictionary.
+        /// </summary>
+        internal PyScope Scope
+        {
+            get { return scope; }
+            set { scope = (PyScope)value; }
+        }
+
+        private static PyScope GlobalScope;
+
+        private PyScope CreateUniquePythonScope()
+        {
+            PyScopeManager.Global.TryGet(uniquePythonScopeName, out PyScope Scope);
+
+            if (Scope == null)
+            {
+                Scope = Py.CreateScope(uniquePythonScopeName);
+                Scope.Exec(@"
+import clr
+import sys
+");
+            }
+
+            return Scope;
+        }
+
+        private static PyScope CreateGlobalScope()
+        {
+            var scope = Py.CreateScope(globalScopeName);
+            // Allows discoverability of modules by inspecting their attributes
+            scope.Exec(@"
+import clr
+clr.setPreload(True)
+");
+
+            return scope;
+        }
+
+        private object ExecutePythonScriptCode(string code)
+        {
+            Python.Included.Installer.SetupPython().Wait();
+
+            if (!PythonEngine.IsInitialized)
+            {
+                PythonEngine.Initialize();
+                PythonEngine.BeginAllowThreads();
+            }
+
+            using (Py.GIL())
+            {
+                var result = Scope.Eval(code);
+                return CPythonEvaluator.OutputMarshaler.Marshal(result);
+            }
+        }
+
+        /// <summary>
+        /// List all of the members in a PythonObject
+        /// </summary>
+        /// <param name="pyObject">A reference to the module</param>
+        /// <param name="name">The name of the module</param>
+        /// <returns>A list of completion data for the module</returns>
+        private IEnumerable<Tuple<string, string, bool, ExternalCodeCompletionType>> EnumerateMembersForPyObj(object pyObject, string name)
+        {
+            var items = new List<Tuple<string, string, bool, ExternalCodeCompletionType>>();
+            var d = pyObject as PyObject;
+            if (d == null)
+            {
+                return items;
+            }
+
+            foreach (var member in d.Dir())
```

Comment:
so one thing that is a bit unclear to callers of this method is that it needs to be wrapped in the GIL block - I think - comment maybe?

Reply:
Added a comment
Also this is a private method.